File: ui/src/api/giftCards.ts
import { loggedInvoke } from '@/utils/logged-invoke';

/** A gift card record with balance and status. */
export interface GiftCard {
  id: string;
  card_number: string;
  pin: string;
  initial_balance_minor: number;
  current_balance_minor: number;
  currency: string;
  status: string;
  issued_to: string;
  issue_date: string;
  expiry_date: string | null;
  created_by: string | null;
  updated_at: string;
}

/** A transaction against a gift card (issue, redeem, top-up, freeze). */
export interface GiftCardTransaction {
  id: string;
  gift_card_id: string;
  sale_id: string | null;
  txn_type: string;
  amount_minor: number;
  balance_after_minor: number;
  notes: string;
  created_at: string;
}

/** A gift card with its full transaction history. */
export interface GiftCardWithTransactions {
  card: GiftCard;
  transactions: GiftCardTransaction[];
}

/** Filter parameters for listing gift cards. */
export interface GiftCardFilter {
  search?: string | null;
  status?: string | null;
  issued_to?: string | null;
  min_balance?: number | null;
}

/** Input for issuing a new gift card. */
export interface IssueGiftCardInput {
  card_number: string;
  pin?: string | null;
  initial_amount_minor: number;
  currency: string;
  issued_to?: string | null;
  created_by: string;
  expiry_date?: string | null;
}

/** Gift card balance check result. */
export interface BalanceResult {
  balance_minor: number;
  currency: string;
  status: string;
}

/** Result of redeeming a gift card against a sale. */
export interface RedeemGiftCardResult {
  card: GiftCard;
  transaction: GiftCardTransaction;
}

/** Issue a new gift card with an initial balance (session-scoped). */
export const issueGiftCard = (sessionToken: string, input: IssueGiftCardInput): Promise<GiftCardWithTransactions> =>
  loggedInvoke<GiftCardWithTransactions>('issue_gift_card_scoped', { sessionToken, input });

/** Get a gift card by card number or ID, including transactions (session-scoped). */
export const getGiftCard = (sessionToken: string, cardNumberOrId: string): Promise<GiftCardWithTransactions | null> =>
  loggedInvoke<GiftCardWithTransactions | null>('get_gift_card_scoped', { sessionToken, cardNumberOrId });

/** List gift cards with optional filtering (session-scoped). */
export const listGiftCards = (sessionToken: string, filter: GiftCardFilter): Promise<GiftCardWithTransactions[]> =>
  loggedInvoke<GiftCardWithTransactions[]>('list_gift_cards_scoped', { sessionToken, filter });

/** Check a gift card's current balance and status (session-scoped). */
export const getGiftCardBalance = (sessionToken: string, cardNumberOrId: string): Promise<BalanceResult | null> =>
  loggedInvoke<BalanceResult | null>('get_gift_card_balance_scoped', { sessionToken, cardNumberOrId });

/** Redeem a gift card for a given amount against a sale (session-scoped). */
export const redeemGiftCard = (sessionToken: string, cardNumberOrId: string, amountMinor: number, saleId: string): Promise<RedeemGiftCardResult> =>
  loggedInvoke<RedeemGiftCardResult>('redeem_gift_card_scoped', { sessionToken, cardNumberOrId, amountMinor, saleId });

/** Add funds to an existing gift card (session-scoped). */
export const topUpGiftCard = (sessionToken: string, cardNumberOrId: string, amountMinor: number): Promise<GiftCardWithTransactions> =>
  loggedInvoke<GiftCardWithTransactions>('top_up_gift_card_scoped', { sessionToken, cardNumberOrId, amountMinor });

/** Freeze a gift card to prevent further use (session-scoped). */
export const freezeGiftCard = (sessionToken: string, cardNumberOrId: string): Promise<GiftCard> =>
  loggedInvoke<GiftCard>('freeze_gift_card_scoped', { sessionToken, cardNumberOrId });

/** Unfreeze a previously frozen gift card (session-scoped). */
export const unfreezeGiftCard = (sessionToken: string, cardNumberOrId: string): Promise<GiftCard> =>
  loggedInvoke<GiftCard>('unfreeze_gift_card_scoped', { sessionToken, cardNumberOrId });

File: apps/tablet-client/src/commands/gift_cards.rs
use serde::Serialize;
use tauri::{State, command};

use oz_core::db::Store;
use oz_core::gift_card::{
    GiftCard, GiftCardFilter, GiftCardWithTransactions, IssueGiftCardInput, RedeemGiftCardResult,
};

use crate::error::AppError;
use crate::state::AppState;

#[derive(Debug, Serialize)]
/// Balanceresult.
pub struct BalanceResult {
    /// Balance Minor.
    pub balance_minor: i64,
    /// ISO-4217 currency code.
    pub currency: String,
    /// Current status.
    pub status: String,
}

#[command]
/// Issue gift card.
pub async fn issue_gift_card(
    input: IssueGiftCardInput,
    state: State<'_, AppState>,
) -> Result<GiftCardWithTransactions, AppError> {
    let db = state.db.lock().await;
    let store = Store::new(&db);
    let result = store.issue_gift_card(input)?;
    drop(db);
    Ok(result)
}

#[command]
/// Get gift card.
pub async fn get_gift_card(
    card_number_or_id: String,
    state: State<'_, AppState>,
) -> Result<Option<GiftCardWithTransactions>, AppError> {
    let db = state.db.lock().await;
    let store = Store::new(&db);
    let result = store.get_gift_card_detail(&card_number_or_id)?;
    drop(db);
    Ok(result)
}

#[command]
/// List gift cards.
pub async fn list_gift_cards(
    filter: GiftCardFilter,
    state: State<'_, AppState>,
) -> Result<Vec<GiftCardWithTransactions>, AppError> {
    let db = state.db.lock().await;
    let store = Store::new(&db);
    let result = store.list_gift_cards(filter)?;
    drop(db);
    Ok(result)
}

#[command]
/// Get gift card balance.
pub async fn get_gift_card_balance(
    card_number_or_id: String,
    state: State<'_, AppState>,
) -> Result<Option<BalanceResult>, AppError> {
    let db = state.db.lock().await;
    let store = Store::new(&db);
    let result = store.get_gift_card_balance(&card_number_or_id)?;
    drop(db);
    Ok(
        result.map(|(balance_minor, currency, status)| BalanceResult {
            balance_minor,
            currency,
            status,
        }),
    )
}

#[command]
/// Redeem gift card.
pub async fn redeem_gift_card(
    card_number_or_id: String,
    amount_minor: i64,
    sale_id: String,
    state: State<'_, AppState>,
) -> Result<RedeemGiftCardResult, AppError> {
    let db = state.db.lock().await;
    let store = Store::new(&db);
    let result = store.redeem_gift_card(&card_number_or_id, amount_minor, &sale_id)?;
    drop(db);
    Ok(result)
}

#[command]
/// Top_up gift card.
pub async fn top_up_gift_card(
    card_number_or_id: String,
    amount_minor: i64,
    state: State<'_, AppState>,
) -> Result<GiftCardWithTransactions, AppError> {
    let db = state.db.lock().await;
    let store = Store::new(&db);
    let result = store.top_up_gift_card(&card_number_or_id, amount_minor)?;
    drop(db);
    Ok(result)
}

#[command]
/// Freeze gift card.
pub async fn freeze_gift_card(
    card_number_or_id: String,
    state: State<'_, AppState>,
) -> Result<GiftCard, AppError> {
    let db = state.db.lock().await;
    let store = Store::new(&db);
    let result = store.freeze_gift_card(&card_number_or_id)?;
    drop(db);
    Ok(result)
}

#[command]
/// Unfreeze gift card.
pub async fn unfreeze_gift_card(
    card_number_or_id: String,
    state: State<'_, AppState>,
) -> Result<GiftCard, AppError> {
    let db = state.db.lock().await;
    let store = Store::new(&db);
    let result = store.unfreeze_gift_card(&card_number_or_id)?;
    drop(db);
    Ok(result)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn balance_result_debug() {
        let r = BalanceResult {
            balance_minor: 50000,
            currency: "IDR".into(),
            status: "active".into(),
        };
        let debug = format!("{:?}", r);
        assert!(debug.contains("50000"));
    }

    #[test]
    fn balance_result_serialize() {
        let r = BalanceResult {
            balance_minor: 100000,
            currency: "USD".into(),
            status: "active".into(),
        };
        let json = serde_json::to_value(&r).unwrap();
        assert_eq!(json["balance_minor"], 100000);
        assert_eq!(json["currency"], "USD");
        assert_eq!(json["status"], "active");
    }

    #[test]
    fn balance_result_zero_balance() {
        let r = BalanceResult {
            balance_minor: 0,
            currency: "USD".into(),
            status: "exhausted".into(),
        };
        let json = serde_json::to_value(&r).unwrap();
        assert_eq!(json["balance_minor"], 0);
        assert_eq!(json["status"], "exhausted");
    }
}
